src/nvim/eval.c

```diff
@@ -15534,6 +15534,64 @@ static void f_split(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   p_cpo = save_cpo;
 }
 
+
+/// "stdpath(type)" function
+static void f_stdpath(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  const char *p = tv_get_string(&argvars[0]);
+  const void *iter = NULL;
+
+  if (strcmp(p, "config") == 0) {
+    rettv->v_type = VAR_STRING;
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGConfigHome);
+  } else if (strcmp(p, "data") == 0) {
+    rettv->v_type = VAR_STRING;
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGDataHome);
+  } else if (strcmp(p, "cache") == 0) {
+    rettv->v_type = VAR_STRING;
+    rettv->vval.v_string = (char_u *)get_xdg_home(kXDGCacheHome);
+  } else if (strcmp(p, "config_dirs") == 0) {
+    list_T *const list = tv_list_alloc();
+    rettv->v_type = VAR_LIST;
+    rettv->vval.v_list = list;
+    list->lv_refcount++;
+    char *const dirs = stdpaths_get_xdg_var(kXDGConfigDirs);
+    do {
+      size_t dir_len;
+      const char *dir;
+      iter = vim_colon_env_iter(dirs, iter, &dir, &dir_len);
+      if (dir != NULL && dir_len > 0) {
+        char *dir_with_nvim = strncpy(xmallocz(dir_len), dir, dir_len);
+        dir_with_nvim = concat_fnames_realloc(dir_with_nvim, "nvim", true);
+        tv_list_append_string(list, dir_with_nvim, strlen(dir_with_nvim));
+        xfree(dir_with_nvim);
+      }
+    } while (iter != NULL);
+    xfree(dirs);
+  } else if (strcmp(p, "data_dirs") == 0) {
+    list_T *const list = tv_list_alloc();
+    rettv->v_type = VAR_LIST;
+    rettv->vval.v_list = list;
+    list->lv_refcount++;
+    char *const dirs = stdpaths_get_xdg_var(kXDGDataDirs);
+    do {
+      size_t dir_len;
+      const char *dir;
+      iter = vim_colon_env_iter(dirs, iter, &dir, &dir_len);
+      if (dir != NULL && dir_len > 0) {
+        char *dir_with_nvim = strncpy(xmallocz(dir_len), dir, dir_len);
+        dir_with_nvim = concat_fnames_realloc(dir_with_nvim, "nvim", true);
+        tv_list_append_string(list, dir_with_nvim, strlen(dir_with_nvim));
+        xfree(dir_with_nvim);
+      }
+    } while (iter != NULL);
+    xfree(dirs);
```

Comment:
Code for `config_dirs` and `data_dirs` cases are nearly identical. Could be refactored.